### PR TITLE
[Routing] Allow force-generation of trailing parameters using eg "/exports/news.{!_format}"

### PR DESCRIPTION
**Pull request by luchaninov**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #29593
| License       | MIT

When a route is defined as path: `/exports/news.{!_format}`, we should force `_format` be defined in `defaults` and the generator should generate URLs with that default when none is provided (should work with any parameter of course).

### CODE REVIEW

**Review by nicolas-grekas — APPROVED** [review]
awesome! just a minor comment and good on my side!

**Review by stloyd — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 1]
```diff
@@ -156,21 +156,32 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
         $message = 'Parameter "{parameter}" for route "{route}" must match "{expected}" ("{given}" given) to generate a corresponding URL.';
         foreach ($tokens as $token) {
             if ('variable' === $token[0]) {
-                if (!$optional || !array_key_exists($token[3], $defaults) || null !== $mergedParams[$token[3]] && (string) $mergedParams[$token[3]] !== (string) $defaults[$token[3]]) {
+                $varName = $token[3];
+                if ('!' === substr($varName, 0, 1)) {
```
[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/Routing/Generator/UrlGenerator.php]: or better: `'!' === $varName[0]`

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 2]
```diff
@@ -156,21 +156,32 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
...
+                    $varName = substr($varName, 1);
+                    $coalescing = true;
+                } else {
+                    $coalescing = false;
```
[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/Routing/Generator/UrlGenerator.php]: or in one line: `if ($important = '!' === $varName[0]) {`

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (luchaninov) [reply src/Symfony/Component/Routing/Generator/UrlGenerator.php]: @stloyd My IDE also suggested changing to strpos but I thought that substr will be faster. 

@nicolas-grekas `$varName[0]` would be awesome but I'm afraid of `PHP Notice:  Uninitialized string offset: 0` if the string happens to be empty.

So I've checked
https://gist.github.com/luchaninov/c8dd1f95aaeb8426f1af3a81c2dda8f6

`$varName[0]` solution wins. Is it safe to use?

[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/Routing/Generator/UrlGenerator.php]: yes, per the regexp below: https://github.com/symfony/symfony/pull/29599/files#diff-aeec0cbcb3837463c63c212f50cd860fR114
the varname cannot be empty

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 3]
```diff
@@ -156,21 +156,32 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
         $message = 'Parameter "{parameter}" for route "{route}" must match "{expected}" ("{given}" given) to generate a corresponding URL.';
         foreach ($tokens as $token) {
             if ('variable' === $token[0]) {
-                if (!$optional || !array_key_exists($token[3], $defaults) || null !== $mergedParams[$token[3]] && (string) $mergedParams[$token[3]] !== (string) $defaults[$token[3]]) {
+                $varName = $token[3];
+                if ('!' === substr($varName, 0, 1)) {
+                    $varName = substr($varName, 1);
+                    $coalescing = true;
```
[the remark this answers is not in the document]
Comment (luchaninov) [reply src/Symfony/Component/Routing/Generator/UrlGenerator.php]: Thanks, that's better. Renamed to `$important`

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 4]
```diff
@@ -156,21 +156,32 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
...
+                } else {
+                    $coalescing = false;
+                }
+
+                if (
```
[the remark this answers is not in the document]
Comment (luchaninov) [reply src/Symfony/Component/Routing/Generator/UrlGenerator.php]: Made condition oneliner

**Comment by luchaninov** [comment]
> Would you mind creating a doc issue/PR?

https://github.com/symfony/symfony-docs/pull/10785

File: src/Symfony/Component/Routing/RouteCompiler.php [hunk 5]
```diff
@@ -184,6 +184,9 @@ private static function compilePattern(Route $route, $pattern, $isHost)
             }
 
             $tokens[] = array('variable', $isSeparator ? $precedingChar : '', $regexp, $varName);
+            if ('!' === substr($varName, 0, 1)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Routing/RouteCompiler.php]: $varName[0] here (and one more below)

File: src/Symfony/Component/Routing/Tests/Generator/UrlGeneratorTest.php [hunk 6]
```diff
@@ -397,6 +397,27 @@ public function testDefaultRequirementOfVariable()
         $this->assertSame('/app.php/index.mobile.html', $generator->generate('test', array('page' => 'index', '_format' => 'mobile.html')));
     }
 
+    public function testCoalescingVariable()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Routing/Tests/Generator/UrlGeneratorTest.php]: testImportantVariable? (same below)

File: src/Symfony/Component/Routing/RouteCompiler.php (hunk 5 above)
Comment (luchaninov) [reply src/Symfony/Component/Routing/RouteCompiler.php]: thanks, fixed

File: src/Symfony/Component/Routing/Tests/Generator/UrlGeneratorTest.php (hunk 6 above)
Comment (luchaninov) [reply src/Symfony/Component/Routing/Tests/Generator/UrlGeneratorTest.php]: thanks, fixed

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (luchaninov) [reply src/Symfony/Component/Routing/Generator/UrlGenerator.php]: @nicolas-grekas nice one. To make it more understandable added `( )`: 
`$important = ('!' === $varName[0])`

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (luchaninov) [reply src/Symfony/Component/Routing/Generator/UrlGenerator.php]: changed to ``$varName[0]``

File: src/Symfony/Component/Routing/Tests/Matcher/UrlMatcherTest.php [hunk 7]
```diff
@@ -177,6 +177,15 @@ public function testMatchSpecialRouteName()
         $this->assertEquals(array('_route' => '$péß^a|'), $matcher->match('/bar'));
     }
 
+    public function testMatchCoalescingVariable()
```
[the remark this answers is not in the document]
Comment (luchaninov) [reply src/Symfony/Component/Routing/Tests/Matcher/UrlMatcherTest.php]: sorry :( 
fixed

**Review by OskarStark — APPROVED** [review]
Nice improvement 👍🏻

**Review by stloyd — APPROVED** [review]

**Comment by Tobion** [comment]
I think the solution with `!` is good for forcing the generation of urls with default params. This is already something I tried to tacle before in #5180.
But I suggest to also change the matching for `!` prefixed placeholders to not make them optional.
So `/news.{!_format}` does not match `/news`. Otherwise we promote duplicate URIs again. If you need `/news` to also match, you just create a redirect route as explained in https://github.com/symfony/symfony/pull/28633#issuecomment-431766878

**Comment by nicolas-grekas** [comment]
See #29763, help wanted.

**Pull request merged**